### PR TITLE
Refactor: typing

### PR DESCRIPTION
**Pull request by Borda**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/functional/classification/average_precision.py [hunk 1]
```diff
@@ -102,5 +102,6 @@ def average_precision(
         [tensor(1.), tensor(1.), tensor(0.2500), tensor(0.2500), tensor(nan)]
 
     """
+    # fixme: `sample_weights` is unused
```
Comment (Borda) [inline torchmetrics/functional/classification/average_precision.py:105]: @SkafteNicki any idea why this is not used?

**Comment by Borda** [comment]
@SkafteNicki @justusschock any idea why the differentiability tests are failing? Cannot reproduce it locally :confused:

File: torchmetrics/functional/classification/auroc.py [hunk 2]
```diff
@@ -99,7 +99,7 @@ def _auroc_compute(
 
             # calculate average
             if average == AverageMethod.NONE:
-                return auc_scores
+                return tensor(auc_scores)
```
Comment (maximsch2) [inline torchmetrics/functional/classification/auroc.py:102]: how did you find this issue if it's not covered by tests?

Comment (Borda) [reply torchmetrics/functional/classification/auroc.py:102]: mypy :]

File: torchmetrics/functional/classification/auroc.py [hunk 3]
```diff
@@ -86,7 +86,7 @@ def _auroc_compute(
             tpr = [o[1] for o in output]
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/auroc.py:86]: Then we should probably have another else here
```suggestion
            tpr = [o[1] for o in output]
        else:
            raise ValueError('Detected input to be `multilabel` but you did not provide `num_classes` argument')
```

File: torchmetrics/functional/classification/dice.py [hunk 4]
```diff
@@ -98,12 +98,12 @@ def dice_score(
 
     """
     num_classes = preds.shape[1]
-    bg = (1 - int(bool(bg)))
-    scores = torch.zeros(num_classes - bg, device=preds.device, dtype=torch.float32)
-    for i in range(bg, num_classes):
+    bg_inv = (1 - int(bool(bg)))
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/dice.py:101]:
```suggestion
    bg_inv = (1 - int(bg))
```

File: torchmetrics/functional/regression/explained_variance.py [hunk 5]
```diff
@@ -40,7 +40,7 @@ def _explained_variance_compute(
     sum_target: Tensor,
     sum_squared_target: Tensor,
     multioutput: str = "uniform_average",
-) -> Union[Tensor, Sequence[Tensor]]:
+) -> Tensor:
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/explained_variance.py:43]: This does not seem correct to me. Even when setting `multioutput='raw_values` it is still a single tensor that is returned

File: torchmetrics/functional/regression/pearson.py [hunk 6]
```diff
@@ -22,7 +22,7 @@
 def _pearson_corrcoef_update(
     preds: Tensor,
     target: Tensor,
-    *_,
+    *_: Any,
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/pearson.py:25]: Should be able to just delete?
```suggestion
```

**Review by SkafteNicki — APPROVED** [review]
Few commets, else LGTM :]

File: torchmetrics/functional/regression/explained_variance.py (hunk 5 above)
Comment (Borda) [reply torchmetrics/functional/regression/explained_variance.py:43]: we return the Tensor now or am I missing anything?

File: torchmetrics/utilities/imports.py [hunk 7]
```diff
@@ -68,11 +68,11 @@ def _compare_version(package: str, op: Callable, version: str) -> Optional[bool]
...
 
-_TORCH_LOWER_1_4 = _compare_version("torch", operator.lt, "1.4.0")
-_TORCH_LOWER_1_5 = _compare_version("torch", operator.lt, "1.5.0")
-_TORCH_LOWER_1_6 = _compare_version("torch", operator.lt, "1.6.0")
-_TORCH_GREATER_EQUAL_1_6 = _compare_version("torch", operator.ge, "1.6.0")
-_TORCH_GREATER_EQUAL_1_7 = _compare_version("torch", operator.ge, "1.7.0")
-_LIGHTNING_AVAILABLE = _module_available("pytorch_lightning")
-_TORCH_FIDELITY_AVAILABLE = _module_available("torch_fidelity")
-_SCIPY_AVAILABLE = _module_available("scipy")
+_TORCH_LOWER_1_4: Optional[bool] = _compare_version("torch", operator.lt, "1.4.0")
```
Comment (maximsch2) [inline torchmetrics/utilities/imports.py:71]: Given that we don't really work without torch, having this as Optional feels weird

Comment (Borda) [reply torchmetrics/utilities/imports.py:71]: this comes from typing...

**Pull request merged**